Lower allocate clause from Flang parse tree to OpenMP MLIR

Rin18 opened the pull request.

kiranchandramohan commented:
@Rin18 did you miss the tests?

SouraVX commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -178,14 +179,27 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+      } else if(const auto &allocateClause = 
+                  std::get_if<Fortran::parser::OmpAllocateClause>(&clause.u)){
+        if (const auto &allocatorValue = 
+              std::get<std::optional<Fortran::parser::OmpAllocateClause::Allocator>>(allocateClause->t)){
```
Unsure of this formatting here. Did you clang-formatted it before submitting ?

Rin18 replied:
No, I missed that, I will do so when I update this review. Thanks for pointing it out.

SouraVX commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -138,9 +138,10 @@ genOMP(Fortran::lower::AbstractConverter &converter,
     SmallVector<Value, 4> privateClauseOperands, firstprivateClauseOperands,
-        sharedClauseOperands, copyinClauseOperands;
+        sharedClauseOperands, copyinClauseOperands, 
+        allocatorOperands, allocateOperands;
```
`allocatorOperands` is `SmallVector`. Can there be multiple allocator within single clause ?

Rin18 replied:
In the Fortran syntax there can not be more than one allocator per allocate clause, however, that is different in the mlid implementation of the allocate clause. In order to keep track of which value have what allocator, I enforced that each value will have an allocator, event if one is not specified(in this case it will be the default allocator). So yes, there will be multiple allocators within a single clause and the allocateOperands SmallVector keeps track of them.

SouraVX commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -178,14 +179,27 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+              std::get<std::optional<Fortran::parser::OmpAllocateClause::Allocator>>(allocateClause->t)){
+          allocator = fir::getBase(converter.genExprValue(*Fortran::semantics::GetExpr(allocatorValue->v)));
+          allocatorOperands.push_back(allocator);
+        } else{
```
Not sure of this piece of code? Could you please elaborate. FWIW you're pushing into the vector ? but again as my previous comment ^above. Do we need a vector in first place.

Rin18 replied:
Here I am checking whether the allocate clause has an allocator specified, in which case I push it into the vector. If it does not have an allocator specified, then I push a default allocator into the vector.

SouraVX reviewed (commented):
Thanks for the patch @Rin18!
Some general comments inline. Please add the test case. IMHO test case can help clarifying/understanding the associated code change much better.

clementval commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -178,14 +179,27 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+                  std::get_if<Fortran::parser::OmpAllocateClause>(&clause.u)){
+        if (const auto &allocatorValue = 
+              std::get<std::optional<Fortran::parser::OmpAllocateClause::Allocator>>(allocateClause->t)){
+          allocator = fir::getBase(converter.genExprValue(*Fortran::semantics::GetExpr(allocatorValue->v)));
```
I think `allocator` is not used after being pushed in the vector. Would it make sense to declare a simple Value here in this scope?

Rin18 replied:
I'll try and do that

clementval commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -178,14 +179,27 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+          allocatorOperands.push_back(allocator);
+        } else{
+          auto val = firOpBuilder.getI32IntegerAttr(1);
+          allocator = firOpBuilder.createIntegerConstant(currentLocation, val.getType(), 1);
```
Same comment for `allocator` here.

clementval commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -178,14 +179,27 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+      } else if(const auto &allocateClause = 
+                  std::get_if<Fortran::parser::OmpAllocateClause>(&clause.u)){
+        if (const auto &allocatorValue = 
+              std::get<std::optional<Fortran::parser::OmpAllocateClause::Allocator>>(allocateClause->t)){
```
You might to check that `allocatorValue` has a value since it is an `std::optional`

Rin18 replied:
Thanks for pointing it out, will do that.

clementval commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -178,14 +179,38 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+        } else {
+          while (count--) {
+            auto val = firOpBuilder.getI32IntegerAttr(1);
+            allocator = firOpBuilder.createIntegerConstant(currentLocation,
```
Can you reuse the same constant for each operand or do you really need to have the same constant created multiple times?

Rin18 replied:
I can reuse the same constant for the val here

Rin18 commented:
Really sorry, but I forgot to squash my commits. I have accidentally pushed 3 of my commits instead of squashing them. They should be responding to the review comments that were left.

clementval commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -178,14 +179,39 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+          allocator = firOpBuilder.createIntegerConstant(currentLocation,
+                                                         val.getType(), 1);
+          while (count--) {
+
```
Remove empty line.

schweitzpgi commented on flang/test/lit.cfg.py:
```diff
@@ -87,6 +87,9 @@
   ToolSubst('%f18', command=FindTool('f18'),
+    extra_args=["-intrinsic-module-directory "+config.flang_intrinsic_modules_dir],
+    unresolved='fatal'),
+  ToolSubst('%bbc', command=FindTool('bbc'),
```
Is this going to break all the tests that use `bbc` and do not specify this option?

Rin18 replied:
It doesn't seem to break them. I checked the tests and they don't break, whether they use this option or not.

schweitzpgi commented:
What's the status on this one, Irina? I don't see any approved checkmarks.

Rin18 commented:
This patch is not approved yet. I have addressed the review comments and I'm now waiting for someone to look at it again and approve it / ask for other changes.

SouraVX commented on flang/test/Lower/OpenMP/omp-parallel-allocate-clause.f90:
```diff
@@ -0,0 +1,29 @@
+!FIRDialect:    omp.terminator
+!FIRDialect:  }
+subroutine allocate_clause(arg1, arg2)
+use omp_lib
```
Is this line strictly needed ? I reckon you can do this without using `use omp_lib` and get away with the change introduced in `flang/test/lit.cfg.py`

Rin18 replied:
I'll try this and see if the test still passes.

kiranchandramohan replied:
Since a predefined value (omp_large_cap_mem_alloc) is used, using the omp_lib is the right way.

SouraVX reviewed (commented):
Thanks for incorporating comments, overall business logic LGTM :) 
Some comments inline.

Rin18 commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -178,14 +179,39 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+        const auto &allocatorValue = std::get<
+            std::optional<Fortran::parser::OmpAllocateClause::Allocator>>(
+            allocateClause->t);
+        if (allocatorValue) {
```
Sure thing, I'll do so.

Rin18 commented on flang/test/Lower/OpenMP/omp-parallel-allocate-clause.f90:
```diff
@@ -0,0 +1,29 @@
+        print*, alpha, beta, gama
+!$OMP END PARALLEL
+
+end subroutine
```
Yeah, I think VS Code does that. I'll make sure to fix this.

Rin18 commented on flang/test/lit.cfg.py:
```diff
@@ -87,6 +87,9 @@
 # the build directory holding that tool.
 tools = [
   ToolSubst('%f18', command=FindTool('f18'),
+    extra_args=["-intrinsic-module-directory "+config.flang_intrinsic_modules_dir],
```
That sounds good, the only problem is that the test I have for this patch would fail without the change in lit.cfg.py. I'm not sure what the best approach would be because of that. Should I have this change in a separate patch, remove the test for this patch and add it later when lit.cfg.py is upstreamed? Is there a better approach?

kiranchandramohan replied:
You can add this entry for bbc in lit.cfg.py as a separate patch in fir-dev and if necessary (the bbc tool is not available upstream) upstream also.

Rin18 replied:
Alright, I'll make a separate patch in fir-dev for this entry

Rin18 commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -178,14 +179,39 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+      } else if (const auto &allocateClause =
+                     std::get_if<Fortran::parser::OmpAllocateClause>(
+                         &clause.u)) {
+        mlir::Value allocator;
```
Sure thing, I'll make sure to change this

Rin18 commented on flang/test/Lower/OpenMP/omp-parallel-allocate-clause.f90:
```diff
@@ -0,0 +1,29 @@
+! RUN:   FileCheck %s --check-prefix=FIRDialect
+
+!FIRDialect: func @_QPallocate_clause(%arg0: !fir.ref<f32>, %arg1: !fir.ref<f32>) {
+!FIRDialect-DAG: %0 = fir.alloca i32 {name = "_QFallocate_clauseEa"}
```
I see, I'll make sure to avoid that in the future and fix this. Thank you!

schweitzpgi replied:
That is a script for automated check generation at `mlir/utils/generate-test-checks.py`.

kiranchandramohan commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -178,14 +179,39 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+        if (allocatorValue) {
+          allocator = fir::getBase(converter.genExprValue(
+              *Fortran::semantics::GetExpr(allocatorValue->v)));
+          while (count--) {
```
Will any of the usage in the following link work here for small vector?
https://stackoverflow.com/questions/30998444/add-same-value-multiple-times-to-stdvector-repeat

Rin18 replied:
Thank you for the link Kiran, I'll make sure to give this a go

kiranchandramohan commented on flang/test/Lower/OpenMP/omp-parallel-allocate-clause.f90:
```diff
@@ -0,0 +1,29 @@
+!FIRDialect-DAG: %6 = fir.alloca i32 {name = "_QFallocate_clauseEy"}
+!FIRDialect-DAG: %7 = fir.alloca i32 {name = "_QFallocate_clauseEz"}
+  
+!FIRDialect-DAG: omp.parallel private(%1 : !fir.ref<i32>, %3 : !fir.ref<i32>, %4 : !fir.ref<i32>) allocate(%c1_i32 : i32 -> %5 : !fir.ref<i32>, %c1_i32 : i32 -> %6 : !fir.ref<i32>, %c1_i32 : i32 -> %7 : !fir.ref<i32>, %c1_i32_0 : i32 -> %0 : !fir.ref<i32>, %c1_i32_0 : i32 -> %2 : !fir.ref<i32>) {
```
Should check that the allocator used is the one specified.

kiranchandramohan reviewed (commented):
A few comments.

Rin18 commented on flang/test/Lower/OpenMP/omp-parallel-allocate-clause.f90:
```diff
@@ -0,0 +1,29 @@
+use omp_lib
+        integer :: alpha, beta, gama, x, y, z, a, b
+
+!$OMP PARALLEL PRIVATE(alpha, beta, gama) ALLOCATE(omp_large_cap_mem_alloc:x, y, z) ALLOCATE(a, b)
```
Yes, there is, I'll update this test.

Rin18 commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -178,14 +179,39 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+          }
+        } else {
+          auto val = firOpBuilder.getI32IntegerAttr(1);
+          allocator = firOpBuilder.createIntegerConstant(currentLocation,
```
I think you're right, I'll update this.

SouraVX commented:
One small thing, since most of the review is done. You may better rebase and squash things up, that would help in having a final look and landing from their.
NIT: This `lit` file changes has to get in first before this.

schweitzpgi commented:
> One small thing, since most of the review is done. You may better rebase and squash things up, that would help in having a final look and landing from their.
> NIT: This `lit` file changes has to get in first before this.

Right. Unfortunately, this one got caught up in the last fir-dev rebase.

Rin18 commented:
> One small thing, since most of the review is done. You may better rebase and squash things up, that would help in having a final look and landing from their.
> NIT: This `lit` file changes has to get in first before this.
Since the 'lit' file changes have been accepted I went ahead and rebased and squashed the commits. Hope everything is in order now.

SouraVX commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -178,14 +179,38 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+        mlir::Value allocatorOperand;
+        const Fortran::parser::OmpObjectList &ompObjectList =
+            std::get<Fortran::parser::OmpObjectList>(allocateClause->t);
+        /// Check if allocate clause has allocator specified. If so, add it
```
NIT: No need for Doxygen style comments ?
Use C++ style instead '//' . And Finish it with period.

Rin18 replied:
I see, I'll change this right away. Thanks for the review

SouraVX reviewed (approved):
LGTM.

kiranchandramohan commented on flang/test/Lower/OpenMP/omp-parallel-allocate-clause.f90:
```diff
@@ -0,0 +1,29 @@
+!FIRDialect-DAG: %[[Y:.*]] = fir.alloca i32 {name = "{{.*}}Ey"}
+!FIRDialect-DAG: %[[Z:.*]] = fir.alloca i32 {name = "{{.*}}Ez"}
+
+!FIRDialect-DAG: %[[LARGE:.*]] = constant 1 : i32
+!FIRDialect-DAG: %[[DEFAULT:.*]] = constant 1 : i32
```
Why are these the same for both default and large mem allocation? Shouldn't they be different constants?

Rin18 replied:
I'm not entirely sure as the constants are taken from the parse tree, except in the "no allocator specified" case, where I pass 1 as the constant.

kiranchandramohan replied:
Ahh OK. I found that the entry of large_cap_mem_alloc is set to default in the omp_lib.h file. We have to either update this file or use the one provided with the OpenMP runtime in flang.

1) Relevant info in Flang omp_lib.f90 (it includes omp_lib.h)
https://github.com/llvm/llvm-project/blob/13042da5cb3fe5f17c58c2446c93635a2c432049/flang/module/omp_lib.h#L106

2) Relevant info in OpenMP runtime omp_lib.f90
https://github.com/llvm/llvm-project/blob/13042da5cb3fe5f17c58c2446c93635a2c432049/openmp/runtime/src/include/omp_lib.f90.var#L132

kiranchandramohan commented on flang/test/Lower/OpenMP/omp-parallel-allocate-clause.f90:
```diff
@@ -0,0 +1,29 @@
+!FIRDialect-DAG: %[[LARGE:.*]] = constant 1 : i32
+!FIRDialect-DAG: %[[DEFAULT:.*]] = constant 1 : i32
+  
+!FIRDialect: omp.parallel private(%[[X]] : !fir.ref<i32>, %[[Y]] : !fir.ref<i32>, %[[Z]] : !fir.ref<i32>, %[[A]] : !fir.ref<i32>, %[[B]] : !fir.ref<i32>) allocate(%[[LARGE]] : i32 -> %[[X]] : !fir.ref<i32>, %[[LARGE]] : i32 -> %[[Y]] : !fir.ref<i32>, %[[LARGE]] : i32 -> %[[Z]] : !fir.ref<i32>, %[[DEFAULT]] : i32 -> %[[A]] : !fir.ref<i32>, %[[DEFAULT]] : i32 -> %[[B]] : !fir.ref<i32>) {
```
Can these be split into multiple lines?

Rin18 replied:
Yes, I will split it in the next update.

kiranchandramohan commented on flang/test/Lower/OpenMP/omp-parallel-allocate-clause.f90:
```diff
@@ -0,0 +1,29 @@
+! This test checks lowering of OpenMP parallel Directive with
```
Do you know why this test failed previously without the corresponding privates for the allocate variables?

Rin18 replied:
Not quite sure as there are semantic checks in place to take care of that.

kiranchandramohan reviewed (commented):
Sorry, this fell out of my radar. We should try to get this merged this week.

Can you add a few more tests?
1) No allocator specified for all.
2) A single allocator specified for all.
3) Two different allocators (excluding the default) specified.

kiranchandramohan commented on flang/test/Lower/OpenMP/omp-parallel-allocate-clause.f90:
```diff
@@ -0,0 +1,68 @@
+!FIRDialect-DAG: %[[Y:.*]] = fir.alloca i32 {name = "{{.*}}Ey"}
+!FIRDialect-DAG: %[[Z:.*]] = fir.alloca i32 {name = "{{.*}}Ez"}
+
+!FIRDialect-DAG: %[[LARGE:.*]] = constant 2 : i32
+!FIRDialect-DAG: %[[DEFAULT:.*]] = constant 1 : i32
+!FIRDialect-DAG: %[[CONSTANT:.*]] = constant 3 : i32
+  
+!FIRDialect: omp.parallel private(%[[X]] : !fir.ref<i32>, %[[Y]] :
+!fir.ref<i32>, %[[Z]] : !fir.ref<i32>, %[[A]] : !fir.ref<i32>, %[[B]] :
+!fir.ref<i32>) allocate(%[[LARGE]] : i32 -> %[[X]] :
+!fir.ref<i32>, %[[LARGE]] : i32 -> %[[Y]] :
+!fir.ref<i32>, %[[LARGE]] : i32 -> %[[Z]] :
+!fir.ref<i32>, %[[DEFAULT]] : i32 -> %[[A]] :
+!fir.ref<i32>, %[[DEFAULT]] : i32 -> %[[B]] :
+!fir.ref<i32>, %[[CONSTANT]] : i32 -> %[[C]] :
+!fir.ref<i32>, %[[CONSTANT]] : i32 -> %[[D]] : !fir.ref<i32>) {
+!FIRDialect:    omp.terminator
+!FIRDialect:  }
+
+!FIRDialect-DAG: %[[DEFAULT2:.*]] = constant 1 : i32
+!FIRDialect: omp.parallel private(%[[E]] : !fir.ref<i32>, %[[F]] :
+!fir.ref<i32>) allocate(%[[DEFAULT2]] : i32 -> %[[E]] :
+!fir.ref<i32>, %[[DEFAULT2]] : i32 -> %[[F]] : !fir.ref<i32>) {
+!FIRDialect:    omp.terminator
+!FIRDialect:  }
+
+!FIRDialect-DAG: %[[LOWLAT:.*]] = constant 5 : i32
+!FIRDialect: omp.parallel private(%[[G]] : !fir.ref<i32>, %[[H]] :
+!fir.ref<i32>) allocate(%[[LOWLAT]] : i32 -> %[[G]] :
+!fir.ref<i32>, %[[LOWLAT]] : i32 -> %[[H]] : !fir.ref<i32>) {
+!FIRDialect:    omp.terminator
+!FIRDialect:  }
```
If the checks are closer to the loops then it will be easier to read.

Rin18 replied:
I'll fix this in the next update.

kiranchandramohan reviewed (commented):
A few more comments.

Rin18 commented on flang/test/Lower/OpenMP/omp-parallel-allocate-clause.f90:
```diff
@@ -0,0 +1,68 @@
+use omp_lib
+        integer :: x, y, z, a, b, c, d, e, f, g, h
+
+!$OMP PARALLEL PRIVATE(x, y, z, a, b) ALLOCATE(omp_large_cap_mem_alloc: x, y, z) ALLOCATE(a, b) ALLOCATE(omp_const_mem_alloc: c, d)
```
Oh yeah, I forgot to add then to private. It's weird that bbc doesn't flag the error. I've tried this and the file shows an error when I run it with flang, but doesn't when I run it with bbc. I'll ass those 2 variables to private in the next update

kiranchandramohan reviewed (approved):
LGTM.
